internal/controller/pathwaysjob_controller.go

```diff
@@ -0,0 +1,663 @@
+		if (c.Type == string(jobsetv1alpha2.JobSetSuspended) || c.Type == string(jobsetv1alpha2.JobSetCompleted) || c.Type == string(jobsetv1alpha2.JobSetFailed)) && c.Status == metav1.ConditionTrue {
+			log.Info("\n\n PathwaysJob: JobSet in TERMINAL STATE", "Condition ", c.Type)
+		}
+		if (c.Type == string(jobsetv1alpha2.JobSetStartupPolicyCompleted) ||
+			c.Type == string(jobsetv1alpha2.JobSetStartupPolicyInProgress)) && c.Status == metav1.ConditionTrue {
+			log.Info("\n\n PathwaysJob: JobSet in TERMINAL STATE", "Condition ", c.Type)
+		}
+	}
+}
+
+// ---------------------- PATHWAYS HELPERS --------------------------
+// Find TPU version from the worker's type (- used to determine Pathways instance_type)
+func constructTPUVersionFromWorkerType(tpuGKEAcceleratorType pathwaysjob.WorkerType) string {
+	// Worker types are already validated in the YAML.
+	return WorkerTypeToTPUVersionMap[string(tpuGKEAcceleratorType)]
+}
+
+// Validate that topology provided is valid for the provided worker type.
+func validateTPUTopologyWithWorkerType(ctx context.Context, tpuGKEAcceleratorType pathwaysjob.WorkerType, topology string) (string, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if slices.Contains(ValidTpuTopologiesMap[string(tpuGKEAcceleratorType)], topology) {
+		return topology, nil
+	} else {
+		log.Info("Invalid topology!!! ", "Worker type ", string(tpuGKEAcceleratorType), " cannot have topology ", topology)
+		return "", fmt.Errorf("invalid TPU topology for worker type")
+	}
+}
+
+// Calculate the number of VMs based on the Topology (- used in completions/parallelisms)
+func calculateVMsFromTopology(topology string) int32 {
+	parts := strings.Split(topology, "x") // Examples - 2x2x4 or 4x4
+	if len(parts) < 2 {
+		return 0
+	}
+	// Calculate the number of chips based on the Topology.
+	chips := 1
+	for _, part := range parts {
+		num, _ := strconv.Atoi(part)
+		chips *= num
+	}
+	vms := 1
+	chipsperVM := 4
+	if chips >= chipsperVM {
+		vms = chips / chipsperVM
+	}
+	return int32(vms)
+}
+
+// Calculate all TPU related information
+func calculateTPUInfo(ctx context.Context, pw *pathwaysjob.PathwaysJob) error {
+	// setting public variable
+	TPUVersion := constructTPUVersionFromWorkerType(pw.Spec.Workers[0].Type)
+	TPUTopology, err := validateTPUTopologyWithWorkerType(ctx, pw.Spec.Workers[0].Type, pw.Spec.Workers[0].Topology)
+	if err != nil {
+		return err
+	}
+	InstanceType = TPUVersion + ":" + TPUTopology
+	NumVMs = calculateVMsFromTopology(pw.Spec.Workers[0].Topology)
+	return nil
+}
+
+// Constructs the Pathways resource manager container spec for the underlying JobSet
+func MakeResourceManagerContainer(pw *pathwaysjob.PathwaysJob, rmJobName string) (*corev1.Container, error) {
+	truth := true
+
+	rmContainerSpec := corev1.Container{
+		Name:            "pathways-rm",
+		Image:           "us-docker.pkg.dev/cloud-tpu-v2-images-dev/pathways/sanitized_server:latest",
```

Comment:
Is this the tagged image?

Reply:
Updated to public images. Thanks for the catch!